### PR TITLE
Keep writing vars_data to socket until done

### LINKED ISSUE #43748: ios_facts: rawunicodeescape codec can't decode bytes in position 4092-4093: truncated \\uXXXX"} with gather_subset: config
##### SUMMARY
`ios_facts` with `gather_subset: config` is returning `{"msg": "'rawunicodeescape' codec can't decode bytes in position 4092-4093: truncated \\uXXXX"}`

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

ios_facts

##### ANSIBLE VERSION

Tested with 2.6.x and 2.7 devel.

```
ansible 2.6.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/johni/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]


ansible 2.7.0.dev0 (pip_umask_doc 9169daa326) last updated 2018/08/01 22:52:09 (GMT +1100)
  config file = /home/johni/upgrade_ios/ansible.cfg
  configured module search path = [u'/home/johni/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/johni/ansible/hacking/ansible/lib/ansible
  executable location = /home/johni/ansible/hacking/ansible/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]

```

##### CONFIGURATION

```
DEFAULT_HOST_LIST(/home/johni/upgrade_ios/ansible.cfg) = [u'/home/johni/upgrade_ios/inventory']

```

##### OS / ENVIRONMENT

* Ubuntu 16.04 LTS (4) and (5)
* RHEL 7

##### STEPS TO REPRODUCE

```yaml
- name: "Demonstrate connecting to switches"
  hosts: switches
  gather_facts: no

  tasks:
    - name: Gather facts (ops)
      ios_facts:
        gather_subset: config
      when: ansible_network_os == 'ios'
    - name: Display some facts
      debug:
        msg: "The hostname is {{ ansible_net_hostname }} and the OS is {{ ansible_net_version }}"
```

##### EXPECTED RESULTS

Would expect to see the hostname and ansible_net_version.  If removing the config from collection:

```
TASK [Display some facts] ****************************************************************************************
ok: [sw1] => {
    "msg": "The hostname is dclabsw03 and the OS is 15.0(2)SE11"
}

```

##### ACTUAL RESULTS

```
TASK [Display some facts] ****************************************************************************************
task path: /home/johni/upgrade_ios/test.yml:13
<192.168.0.222> attempting to start connection
<192.168.0.222> using connection plugin network_cli
<192.168.0.222>
The full traceback is:
Traceback (most recent call last):
  File "/home/johni/ansible/hacking/ansible/bin/ansible-connection", line 225, in main
    variables = cPickle.loads(vars_data)
UnicodeDecodeError: 'rawunicodeescape' codec can't decode bytes in position 4092-4093: truncated \uXXXX

fatal: [sw1]: FAILED! => {
    "msg": "'rawunicodeescape' codec can't decode bytes in position 4092-4093: truncated \\uXXXX"
}

```

### PR DESCRIPTION
**Pull request by Qalthos**
##### SUMMARY

Fixes #43748, or at least I hope it does

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

ansible-connection

##### ANSIBLE VERSION

```
2.7
```

### CODE REVIEW

File: lib/ansible/executor/task_executor.py [hunk 1]
```diff
@@ -928,14 +928,20 @@ def find_file_in_path(filename):
         # Also need to force a protocol that excludes certain control chars as
         # stdin in this case is a pty and control chars will cause problems.
         # that means only protocol=0 will work.
-        src = cPickle.dumps(self._play_context.serialize(), protocol=0)
-        stdin.write(src)
+        def write_to_socket(obj):
```
Comment (sivel) [inline lib/ansible/executor/task_executor.py]: As a note, we should discourage addition of more nested functions.  Nested functions make testing much harder in the end.

Comment (Qalthos) [reply lib/ansible/executor/task_executor.py]: I can take it out and make it standalone

Comment (abadger) [reply lib/ansible/executor/task_executor.py]: +1 to this comment.  Move it to a higher level, private function or method

**Comment by Qalthos** [comment]
This is not the solution, unfortunately.

**Comment by abadger** [comment]
Truncation does seem like the likely culprit but it looks like this isn't quite the problem.  @Qalthos and I debugged further on slack and came up with sending more than 4096 bytes being the problem.  I came up with this test code to explore a solution: https://gist.github.com/abadger/3c65c3b2d0a8e776bc2b35cd6b45384f

I think there's three broad differences between my test code and what's currently in Ansible that will fix this once they're applied:

* Change ansible-connection to use os.read(stdin.fileno()) instead of stdin.readline().
* Change task_executor to remove canonical mode from the pty 
* Change task_executor to use os.write() to write to the pty.

**Pull request closed without merging**